KEYCLOAK-847 Add support for step up authentication

CorneliaLahnsteiner opened the pull request:
This PR adds support for the step up mechanism in Keycloak, which allows access to clients or resources based on an certain authentication level of a user. The implementation builds on Keycloak's design proposal [1]. 

What has been implemented (for OIDC):
* Text mapping in the client configuration to do the text to value translation
* Putting the achieved level of authentication at the 'acr' claim in both the access token and the id token (either as an integer or a textual representation)
* Condition Execution to execute the flow (Condition 1, Condition 2, ...)
* Set Loa on Success Execution (with the option to store the achieved LoA into the user session or not) - Removed during the PR.
* Extension to the Default Authentication Flow in order to determine if a certain level of authentication is reached

Please refer to the documentation for more information, which is available under [2]. 

[1] - https://github.com/keycloak/keycloak-community/blob/master/design/multi-factor-admin-and-step-up.md
[2] - https://github.com/keycloak/keycloak-documentation/pull/1284

pedroigor commented on services/src/main/java/org/keycloak/authentication/AuthenticationSelectionResolver.java:
```diff
@@ -193,7 +193,7 @@ private static boolean addAllExecutionsFromSubflow(AuthenticationProcessor proce
                 // For conditional execution, we must check if condition is true. Otherwise return false, which means trying next
                 // requiredExecution in the list
-                return !flow.isConditionalSubflowDisabled(ex);
+                return !flow.isConditionalSubflowDisabled(ex, false);
```
Could you please elaborate on why not track the matching result when evaluating conditional authenticators?

romge replied:
When the selection page for authenticators is shown, the results must not be cached because the outcome of the condition may be different during the authenticaion flow process. That is because the level of authentication changes during the flow and must be reevaluated in the conditions.

pedroigor replied:
> It' true that a callback method would make the implementation simpler. As this would be a change to the core ConditionalAuthenticator interface and its processing, we would prefer it, if this could be implemented by the main keycloak developers.

I'm good either way. Let's see what @mposolda thinks about it.

pedroigor commented on services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java:
```diff
@@ -673,10 +677,7 @@ protected AccessToken initToken(RealmModel realm, ClientModel client, UserModel
-        // TODO: Add better acr support. See KEYCLOAK-3314
-        String acr = (AuthenticationManager.isSSOAuthentication(clientSession)) ? "0" : "1";
-        token.setAcr(acr);
+        token.setAcr(getAcr(clientSession));
```
Wouldn't be more natural to map `acr` using a protocol mapper? Possibly having a default client scope for this purpose.

pedroigor replied:
Shall we set the `acr` only if requested either by `claims` or `acr_values`?

romge replied:
We implemented that according to the already existing code. If you want to do it as a protocol mapper, that's ok for us.
The same applies to the question whether to set the acr value only when requested or not. This could also be made configurable in the protocol mapper.

pedroigor replied:
Sounds good to me.

pedroigor reviewed (commented):
@CorneliaLahnsteiner Thanks. Some initial comments after a quick look at your changes.

I'm going to continue looking at it in more detail next week.

Btw, could you please rebase?

cgeorgilakis commented on services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java:
```diff
@@ -693,6 +694,29 @@ protected AccessToken initToken(RealmModel realm, ClientModel client, UserModel
+            if (acr == null) {
+                acr = AcrUtils.mapLoaToAcr(loa, acrLoaMap, acrLoaMap.keySet());
+                if (acr == null) {
+                    acr = String.valueOf(loa);
```
[OIDC core specification](https://openid.net/specs/openid-connect-core-1_0.html#IDToken) mentions about acr:

> An absolute URI or an RFC 6711 [RFC6711] registered name SHOULD be used as the acr value;

We think that it would be more clear if we follow this suggestion.
As far as the implementation, we realm admin will be able to map loa value to a default URI. This default value will be returned to client if Keycloak can not map Loa to Acr.

Do you aggree?

pedroigor replied:
Implementation-wise, the logic here should be enough to support that as long as the admin has bound a LoA with a value, right?

W.r.t to the recommendation, we could make this more explicit in docs.

The logic here is also considering setting `0` in case the user is re-authenticating (cookie-based auth).

cgeorgilakis replied:
Yes. If default value is not set, loa will be returned.
Default URI value for 0 can also be configured in the same way as other LOAs.

pedroigor commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/AuthorizationEndpoint.java:
```diff
@@ -479,6 +482,25 @@ private void updateAuthenticationSession() {
+                return loa == null ? Integer.parseInt(acr) : loa;
+            } catch (NumberFormatException e) {
+                // this is an unknown acr, we assume it is very high
+                return Constants.MAXIMUM_LOA;
```
@cgeorgilakis I agree. It should fail and we can not assume the LoA. I think I understand the reasoning about setting MAX, but looks more correct to just fail because the requirement can not be met.

cgeorgilakis replied:
@pedroigor In order to be clear. It should fail only if claims with essential='true' are all unknown. Otherwise, it should omit this claim value.

I do not know when it will fail. I believe that this is implementation decision.

CorneliaLahnsteiner replied:
@cgeorgilakis @pedroigor Thanks for the review - we really appreciate your valuable feedback!

If the requested acr value is unknown, we return the MAXIMUM_LOA (maximum integer) and this results in an authentication error because the user can never reach this LoA. The authentication error only occurs if the claim is essential="true". Otherwise, we omit an unknown requested acr value. 
As already suspected, the error only occurs after the user has provided the credentials.
Additionally, if an authorization request contains more than one (known) acr value, the minimum of the known acr values is processed.

Sorry for the late reply. We are currently going live with one of our services and therefore our resources are quite blocked, so thanks in advance for you patience. Hence, we are not yet able to answer all comments, but we will try our best to do so in the next week. We will rebase this branch by tomorrow.

cgeorgilakis replied:
@CorneliaLahnsteiner you are right about essential="true". This is the most important.

However, think about the following situation:

> Flow Configuration is same as you descipted in .md file( 1st -> username-password, 2nd -> otp). 
Admin have not set acr value in oidc client.
User is already login with level 1.
Client send only acr 'silver' without essential attribute in claim. Keycloak will set REQUESTED_LEVEL_OF_AUTHENTICATION to highest. So, user will ask to authedicate with 2nd LOA ( otp form) and will return 'silver' ( or 2) as acr.
Documentation says '  If the Claim is not Essential and a requested value cannot be provided, the Authorization Server SHOULD return the session's current acr as the value of the acr Claim.'

It would be great if keycloak in this situation do not ask user to authedicate with otp and return 1 as acr.
However, this issue is little importance. The most important for our team is Keycloak being able to support step authedication based on this PR. Default behaviour should be written in Keycloak documentation.

pedroigor replied:
> @pedroigor In order to be clear. It should fail only if claims with essential='true' are all unknown. Otherwise, it should omit this claim value.

Looks like that the changes here fail even for voluntary claims? I think it makes more sense because in either case (essential or voluntary) they are unknown.

If the current behavior is like that. It is OK for me.

pedroigor replied:
>It would be great if keycloak in this situation do not ask user to authedicate with otp and return 1 as acr.
However, this issue is little importance. The most important for our team is Keycloak being able to support step authedication based on this PR. Default behaviour should be written in Keycloak documentation.

It works for me. I'm missing why a client would want to send `silver`, as in this case, as a voluntary claim if the AS should ignore and send the current `acr`.

romge replied:
If essential="true" and all ACRs are unknown, we could fail directly, but it is not clear to me how to implement it with a useful error page in this method.
In the case of an optional ACR where all provided values are unknown, the server will process all authenticators to reach the MAXIMUM_LOA. But instead to fail, it will return the final reached acr value. The ClientNote FORCE_LEVEL_OF_AUTHENTICATION decides in the end wheter to fail or return the final ACR.

CorneliaLahnsteiner replied:
@cgeorgilakis Once the code review is complete and we have a final draft of the code, we will of course provide detailed documentation on Step up Authentication.

cgeorgilakis replied:
> If essential="true" and all ACRs are unknown, we could fail directly, but it is not clear to me how to implement it with a useful error page in this method. In the case of an optional ACR where all provided values are unknown, the server will process all authenticators to reach the MAXIMUM_LOA. But instead to fail, it will return the final reached acr value. The ClientNote FORCE_LEVEL_OF_AUTHENTICATION decides in the end wheter to fail or return the final ACR.

Thanks for all work on this needed future.
However, implementation for optional ACR not defined in Keycloak client is not following OIDC Specification proposal.

> If the Claim is not Essential and a requested value cannot be provided, the Authorization Server SHOULD return the session's current acr as the value of the acr Claim.

I believe that change is easy in order to be allign with OIDC specification. 
Moreover, it Keycloak documentation should provide information about unknown acr.

CorneliaLahnsteiner replied:
@cgeorgilakis We fixed this behavior for a not essential and unknown acr claim: https://github.com/keycloak/keycloak/pull/7897/commits/4081d8151d1c023be1366e563d7487bed2e19b52#diff-20c882d1eedd43d45e8336221b029e174ff2dc5acf324092b7acdc1cebb98936R310

In detail: Until now, an unknown acr claim was always treated with a maximum LoA, which means that one had to authenticate with username+password and OTP - no matter if the acr claim is essential or not. We have adjusted this behaviour so that in the case of an essential (unknown) acr claim, the maximum LoA is maintained, but in the case of a non-essential (unknown) acr claim, the user is only asked to present one authentication factor instead of both (if he has no session yet). Consequently, the access token and ID token will include the acr of 1 after an successful authentication.

CorneliaLahnsteiner replied:
@cgeorgilakis I will also try my best to adapt the documentation to highlight the different possibilities (essential vs. non-essential claims and unknown claims).

cgeorgilakis replied:
Nice!

pedroigor commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/LevelOfAssuranceFlowTest.java:
```diff
@@ -0,0 +1,227 @@
+ * @author <a href="mailto:sebastian.zoescher@prime-sign.com">Sebastian Zoescher</a>
+ */
+@EnableFeature(value = Profile.Feature.WEB_AUTHN, skipRestart = true, onlyForProduct = true)
+public class LevelOfAssuranceFlowTest extends AbstractTestRealmKeycloakTest {
```
This test is very useful. But I think we should have some more such as:

* Check LoA mapping and how `acr` claim is set in the ID Token
* Check the behavior when the user is authenticated and the client requires a different LoA (step-up).

Also, why we are using `EnableFeature` here and using webauthn feature ?

CorneliaLahnsteiner replied:
@pedroigor We wanted to provide a few integration tests for the PR, but these can definitely be extended further. Our intention was to provide more integration tests after the code review, when we have a final draft of the code.

And we will remove the EnableFeature annotation as it is not needed.

romge replied:
I removed the EnableFeature annotation.

pedroigor reviewed (commented):
In regards to new authenticators, shall we have this capability as a tech preview feature?

pedroigor commented on quarkus/runtime/src/test/java/org/keycloak/provider/quarkus/ConfigurationTest.java:
```diff
@@ -184,7 +185,7 @@ public void testDatabaseDefaults() {
         SmallRyeConfig config = createConfig();
         assertEquals(QuarkusH2Dialect.class.getName(), config.getConfigValue("quarkus.hibernate-orm.dialect").getValue());
-        assertEquals("jdbc:h2:file:~/data/keycloakdb;;AUTO_SERVER=TRUE", config.getConfigValue("quarkus.datasource.jdbc.url").getValue());
+        assertEquals("jdbc:h2:file:" + Paths.get("~", "data", "keycloakdb") + ";;AUTO_SERVER=TRUE", config.getConfigValue("quarkus.datasource.jdbc.url").getValue());
```
Do we need this change?

romge replied:
Otherwise the test fails on Windows.

pedroigor commented on services/src/main/java/org/keycloak/authentication/AuthenticatorUtil.java:
```diff
@@ -0,0 +1,46 @@
+
+    public static int getCurrentLevelOfAuthentication(AuthenticationSessionModel authSession) {
+        String authSessionLoaNote = authSession.getAuthNote(Constants.LEVEL_OF_AUTHENTICATION);
+        String userSessionLoaNote = authSession.getUserSessionNotes().get(Constants.LEVEL_OF_AUTHENTICATION);
```
Perhaps `getOrDefault`. We could also add a `getAuthNotOrDefault` too. It can be useful in other places.

romge replied:
A `getAuthNoteOrDefault` method would be helpful indeed, but we think that should be done as a separate feature request.

pedroigor commented:
@mposolda Would make sense to have a callback method on conditional authenticators to perform additional steps after the flow is executed?

In this particular case, we could potentially avoid adding the `SetLoaAuthenticator`. This should improve UX a lot as you don't have to add two executions and configure both.

romge commented:
> 
> 
> @mposolda Would make sense to have a callback method on conditional authenticators to perform additional steps after the flow is executed?
> 
> In this particular case, we could potentially avoid adding the `SetLoaAuthenticator`. This should improve UX a lot as you don't have to add two executions and configure both.

It' true that a callback method would make the implementation simpler. As this would be a change to the core ConditionalAuthenticator interface and its processing, we would prefer it, if this could be implemented by the main keycloak developers.

romge commented on docs/README_Step up mechanism.md:
```diff
@@ -0,0 +1,56 @@
+
+### No ACR Value Specified in the Request
+The design proposal describes that if no LoA is configured at the client, and in the auth request no LoA (acr_value) is specified, the value -1 shall be used. -1 is a constant that indicates that a full authentication should be performed. 
+This means that Keycloak will always do a full authentication, even if the user is already authenticated to a certain level of authentication, e.g.
```
The assumption here is that the client has no LoA configured and also no Loa (ACR) is requested. That means that it is the same case as if the whole feature has not been implemented (current keycloak state). That's why it is also a question of backwards compatibility. Currently keycloak lets you sign in without reauthentication if you have already a valid session (single sign on). This would be broken when the implementation follows the design proposal as it is stated here, because a full reauthentication would be demanded in that case.
I also don't understand your statement that the LoA in the session can't be trusted, because keycloak currently already trusts the session when reauthenticating users. That is the main purpose of the whole single sign on stuff.

Please correct me if I misunderstood you or overlooked something.

pedroigor replied:
What I meant is that we should use `0` when doing cookie-based auth (re-authentication). Even though the session was previously granted with an LoA > 0.

For instance, I'm wondering about a situation where the user reached a high level of trust and after a while, he/she is redirected back to re-authenticate. Does not looks correct to keep the same level of trust because you are not attesting the end-user identity but only re-authenticating. In this case, LoA should be lowered to `0` and the user seamlessly authenticated.

CorneliaLahnsteiner replied:
@pedroigor That's a good point. When re-authentication with a cookie the acr value should be `0`. 

In connection with the cookie issue, we noticed something else:  counter for failed login attempts, which can be activated when using the brute force feature, is be reset by a successful (cookie) authentication. This is very critical from a security perspective when setting up an authentication flow, that has cookie or username-password authentication for level 1 (acr -> 1) and TOTP (acr ->2) for level 2, as the failed login attempts are reset after a successful authentication (in this authentication flow, solely a cookie is sufficient). An attacker would have endless attempts to guess the TOTP, as they can reset the failed login attempts by simply authenticating the user with the cookie (requesting LoA 1). Is this something you are also concerned about, or is it up to the admin configuring the authentication flow? The first solution that came up is to not reset the failed login attempts with a cookie authentication. We created an Jira issue for that: KEYCLOAK-18840. Thanks!

romge replied:
A solution could be to return LoA of 0 for re-authentication and to start with the LoA from the user session when doing a step up authentication. This way the user has not to re-enter already provided credentials.

CorneliaLahnsteiner replied:
@mposolda @pedroigor: We have changed this behavior to the solution presented by @romge, i.e. we always return 0 when re-authenticating via cookie. This also makes us aligned with the OIDC specification.

pedroigor commented:
@mposolda Shall we re-take this one and make it happen?

mposolda reviewed (changes requested):
@CorneliaLahnsteiner @romge Thanks for the PR! And sorry for the late review.

I've added few comments inline. Some of them are quite minor changes and for some others, it will be good to receive some feedback (CC @pedroigor @cgeorgilakis ).

Besides proposed changes, I am working on small refactoring of this PR to simplify the flow logic and add the callback as proposed by @pedroigor and agreed by @romge . I hope to have this change available today/tomorrow and will ping here once done with the branch including the changes.

mposolda commented on services/src/main/java/org/keycloak/authentication/authenticators/browser/SetLoaAuthenticatorFactory.java:
```diff
@@ -0,0 +1,93 @@
+      .type(ProviderConfigProperty.STRING_TYPE)
+      .add()
+      .property()
+      .name(SetLoaAuthenticator.STORE_IN_USER_SESSION)
```
@CorneliaLahnsteiner ok, Thanks for clarification. It makes sense. I will re-add this option to my commit and will ping you afterwards, so you can use latest changes.

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/utils/AcrUtils.java:
```diff
@@ -0,0 +1,89 @@
+    }
+    if (claimsParam != null) {
+      try {
+        JsonNode claims = JsonSerialization.readValue(claimsParam, JsonNode.class);
```
@CorneliaLahnsteiner yeah, I see. I will try to introduce representation object in my commit. Hopefully I can do it today. But agree this is not a blocker for this PR.

mposolda commented:
> > > +1, I will try to refactor this a bit this way with using of a callback. I will ping once finished.
> > > I hope for later today or tomorrow. One thing is, that callback does not need to be specific to "ConditionalAuthenticator" . I think that any "Authenticator" can in theory register callback at the end of the flow and it does not need to be specific to "Condition" IMO.
> > 
> > 
> > @pedroigor @CorneliaLahnsteiner @romge I've did some changes in my branch https://github.com/mposolda/keycloak/tree/KEYCLOAK-847-step-up-auth. Everything is in the last commit [mposolda@e5c4603](https://github.com/mposolda/keycloak/commit/e5c4603a41757dc255d8a8ae958d0b8f34339ee6) . Summary of the changes:
> > 
> > * Added support for the callbacks in the authentication flow (AuthenticationFlowCallback class) . This is triggered after the flow is successfully finished. We can possibly extend with other callback types (EG. when the flow fails), but I did not yet added it. We can do that if there is request for that.
> > * This allows to remove SetLoaAuthenticator as the condition in the subflow is sufficient to check the condition at the beginning of the flow and also update the level after subflow is successfully finished
> > * The condition does not have option STORE_IN_USER_SESSION. As pointed in previous comment above, I think it may be always good to store the LoA in the userSession (I am not see much reasons to not store it). IMO one less flag is always good :-) Or maybe I am missing where this can be useful? Any use-case for this flag?
> > * Changes in ConditionalLoaAuthenticator and ConditionalLoaAuthenticatorFactory classes are due the formatting as it used the wrong formatting according to Keycloak codebase (Keycloak uses 4 spaces alignment when these classes used 2 spaces). There are almost no changes despite of the fact that there is method "onParentFlowSuccess", which is the callback method triggered at the end of the flow (This is what was previously done by SetLoaAuthenticator before this change)
> > * I've did small refactoring of the DefaultAuthenticationFlow to add support for callbacks. On the other hand, I removed support for "LoA" related logic from the DefaultAuthenticationFlow itself. I think it will be good if this class is not tightly coupled with the "LoA" related logic and IMO it is not needed at all. The tests are passing and also did some manual testing and things are fine.
> > 
> > Do you please have some feedback to these changes? If it works for you, I am fine if you just cherry-pick that commit to this PR. Besides some other things I proposed yesterday, it will be also needed to rebase this PR on top of latest master.
> > But hopefully we are close to have this finally available in latest Keycloak master!
> 
> @mposolda Thank you very much for your feedback and your changes - we really appreciate it. I am currently in the process of reviewing them. Regarding `STORE_IN_USER_SESSION`: I have added my feedback in [#7897 (comment)](https://github.com/keycloak/keycloak/pull/7897#discussion_r737628983).
> 
> I will list here all the tasks to be done from our side (the list will be completed tomorrow):
> 
>     * Add Keycloak documentation and remove ReadMe File
> 
>     * Rebase current master
> 
>     * Provide more integration tests:
>       ** Check LoA mapping and how acr claim is set in the ID Token
>       ** Check the behavior when the user is authenticated and the client requires a different LoA (step-up). This should also involve checking whether the user goes only through the missing steps to achieve a determined LoA.

@CorneliaLahnsteiner Thanks for the clarification. If you agree, I would also consider another task before rebase to latest master - cherry pick the commit from my branch with latest changes described above :-) I think it will be good to have these changes in this PR directly as it affects also documentation, tests etc (SetLoaAuthenticator won't be required).

mposolda commented:
> > > > +1, I will try to refactor this a bit this way with using of a callback. I will ping once finished.
> > > > I hope for later today or tomorrow. One thing is, that callback does not need to be specific to "ConditionalAuthenticator" . I think that any "Authenticator" can in theory register callback at the end of the flow and it does not need to be specific to "Condition" IMO.
> > > 
> > > 
> > > @pedroigor @CorneliaLahnsteiner @romge I've did some changes in my branch https://github.com/mposolda/keycloak/tree/KEYCLOAK-847-step-up-auth. Everything is in the last commit [mposolda@e5c4603](https://github.com/mposolda/keycloak/commit/e5c4603a41757dc255d8a8ae958d0b8f34339ee6) . Summary of the changes:
> > > 
> > > * Added support for the callbacks in the authentication flow (AuthenticationFlowCallback class) . This is triggered after the flow is successfully finished. We can possibly extend with other callback types (EG. when the flow fails), but I did not yet added it. We can do that if there is request for that.
> > > * This allows to remove SetLoaAuthenticator as the condition in the subflow is sufficient to check the condition at the beginning of the flow and also update the level after subflow is successfully finished
> > > * The condition does not have option STORE_IN_USER_SESSION. As pointed in previous comment above, I think it may be always good to store the LoA in the userSession (I am not see much reasons to not store it). IMO one less flag is always good :-) Or maybe I am missing where this can be useful? Any use-case for this flag?
> > > * Changes in ConditionalLoaAuthenticator and ConditionalLoaAuthenticatorFactory classes are due the formatting as it used the wrong formatting according to Keycloak codebase (Keycloak uses 4 spaces alignment when these classes used 2 spaces). There are almost no changes despite of the fact that there is method "onParentFlowSuccess", which is the callback method triggered at the end of the flow (This is what was previously done by SetLoaAuthenticator before this change)
> > > * I've did small refactoring of the DefaultAuthenticationFlow to add support for callbacks. On the other hand, I removed support for "LoA" related logic from the DefaultAuthenticationFlow itself. I think it will be good if this class is not tightly coupled with the "LoA" related logic and IMO it is not needed at all. The tests are passing and also did some manual testing and things are fine.
> > > 
> > > Do you please have some feedback to these changes? If it works for you, I am fine if you just cherry-pick that commit to this PR. Besides some other things I proposed yesterday, it will be also needed to rebase this PR on top of latest master.
> > > But hopefully we are close to have this finally available in latest Keycloak master!
> > 
> > 
> > @mposolda Thank you very much for your feedback and your changes - we really appreciate it. I am currently in the process of reviewing them. Regarding `STORE_IN_USER_SESSION`: I have added my feedback in [#7897 (comment)](https://github.com/keycloak/keycloak/pull/7897#discussion_r737628983).
> > I will list here all the tasks to be done from our side (the list will be completed tomorrow):
> > ```
> > * Add Keycloak documentation and remove ReadMe File
> > 
> > * Rebase current master
> > 
> > * Provide more integration tests:
> >   ** Check LoA mapping and how acr claim is set in the ID Token
> >   ** Check the behavior when the user is authenticated and the client requires a different LoA (step-up). This should also involve checking whether the user goes only through the missing steps to achieve a determined LoA.
> > ```
> 
> @CorneliaLahnsteiner Thanks for the clarification. If you agree, I would also consider another task before rebase to latest master - cherry pick the commit from my branch with latest changes described above :-) I think it will be good to have these changes in this PR directly as it affects also documentation, tests etc (SetLoaAuthenticator won't be required).
> 
> I will update my commit now to re-add the option for STORE_IN_USER_SESSION and possibly add ClaimsRepresentation.

I've updated the branch https://github.com/mposolda/keycloak/tree/KEYCLOAK-847-step-up-auth with last commit https://github.com/mposolda/keycloak/commit/6356bf3b7b2f8aaa98492a216bcba5b07ac2a44d . In addition to the things described above, it re-adds the STORE_IN_USER_SESSION and addresses the processing of claims parameter to avoid manual JSON parsing. Is it ok to cherry pick that commit to this PR?

CorneliaLahnsteiner commented:
> > > > > +1, I will try to refactor this a bit this way with using of a callback. I will ping once finished.
> > > > > I hope for later today or tomorrow. One thing is, that callback does not need to be specific to "ConditionalAuthenticator" . I think that any "Authenticator" can in theory register callback at the end of the flow and it does not need to be specific to "Condition" IMO.
> > > > 
> > > > 
> > > > @pedroigor @CorneliaLahnsteiner @romge I've did some changes in my branch https://github.com/mposolda/keycloak/tree/KEYCLOAK-847-step-up-auth. Everything is in the last commit [mposolda@e5c4603](https://github.com/mposolda/keycloak/commit/e5c4603a41757dc255d8a8ae958d0b8f34339ee6) . Summary of the changes:
> > > > 
> > > > * Added support for the callbacks in the authentication flow (AuthenticationFlowCallback class) . This is triggered after the flow is successfully finished. We can possibly extend with other callback types (EG. when the flow fails), but I did not yet added it. We can do that if there is request for that.
> > > > * This allows to remove SetLoaAuthenticator as the condition in the subflow is sufficient to check the condition at the beginning of the flow and also update the level after subflow is successfully finished
> > > > * The condition does not have option STORE_IN_USER_SESSION. As pointed in previous comment above, I think it may be always good to store the LoA in the userSession (I am not see much reasons to not store it). IMO one less flag is always good :-) Or maybe I am missing where this can be useful? Any use-case for this flag?
> > > > * Changes in ConditionalLoaAuthenticator and ConditionalLoaAuthenticatorFactory classes are due the formatting as it used the wrong formatting according to Keycloak codebase (Keycloak uses 4 spaces alignment when these classes used 2 spaces). There are almost no changes despite of the fact that there is method "onParentFlowSuccess", which is the callback method triggered at the end of the flow (This is what was previously done by SetLoaAuthenticator before this change)
> > > > * I've did small refactoring of the DefaultAuthenticationFlow to add support for callbacks. On the other hand, I removed support for "LoA" related logic from the DefaultAuthenticationFlow itself. I think it will be good if this class is not tightly coupled with the "LoA" related logic and IMO it is not needed at all. The tests are passing and also did some manual testing and things are fine.
> > > > 
> > > > Do you please have some feedback to these changes? If it works for you, I am fine if you just cherry-pick that commit to this PR. Besides some other things I proposed yesterday, it will be also needed to rebase this PR on top of latest master.
> > > > But hopefully we are close to have this finally available in latest Keycloak master!
> > > 
> > > 
> > > @mposolda Thank you very much for your feedback and your changes - we really appreciate it. I am currently in the process of reviewing them. Regarding `STORE_IN_USER_SESSION`: I have added my feedback in [#7897 (comment)](https://github.com/keycloak/keycloak/pull/7897#discussion_r737628983).
> > > I will list here all the tasks to be done from our side (the list will be completed tomorrow):
> > > ```
> > > * Add Keycloak documentation and remove ReadMe File
> > > 
> > > * Rebase current master
> > > 
> > > * Provide more integration tests:
> > >   ** Check LoA mapping and how acr claim is set in the ID Token
> > >   ** Check the behavior when the user is authenticated and the client requires a different LoA (step-up). This should also involve checking whether the user goes only through the missing steps to achieve a determined LoA.
> > > ```
> > 
> > 
> > @CorneliaLahnsteiner Thanks for the clarification. If you agree, I would also consider another task before rebase to latest master - cherry pick the commit from my branch with latest changes described above :-) I think it will be good to have these changes in this PR directly as it affects also documentation, tests etc (SetLoaAuthenticator won't be required).
> > I will update my commit now to re-add the option for STORE_IN_USER_SESSION and possibly add ClaimsRepresentation.
> 
> I've updated the branch https://github.com/mposolda/keycloak/tree/KEYCLOAK-847-step-up-auth with last commit [mposolda@6356bf3](https://github.com/mposolda/keycloak/commit/6356bf3b7b2f8aaa98492a216bcba5b07ac2a44d) . In addition to the things described above, it re-adds the STORE_IN_USER_SESSION and addresses the processing of claims parameter to avoid manual JSON parsing. Is it ok to cherry pick that commit to this PR?

@mposolda Thanks for your changes. We will cherry-pick them to this PR and rebase this branch to the latest master afterwards.

mposolda commented:
@CorneliaLahnsteiner @romge Thanks!

CorneliaLahnsteiner commented on docs/README_Step up mechanism.md:
```diff
@@ -0,0 +1,56 @@
+# Step up Authentiation
```
@mposolda I have completed the documentation regarding the step-up mechanism. I have created a PR for the keycloak-documentation: https://github.com/keycloak/keycloak-documentation/pull/1284.

mposolda commented:
@CorneliaLahnsteiner @romge Thanks for all your work! From my point of view, this PR looks good now!

@pedroigor Do you have a chance to review this? I hope we can have this in KC 16, so hopefully we can address some more improvements as a follow-up tasks (EG. protocol mapper and client scopes for "acr" claim). But if you think there is any blocker, please comment.

Summary of the remaining things for this PR:
- Review of the documentation PR (I've added the review, but waiting for our documentation guy Andy if he has further comments)
- I will try to run our internal pipeline with these changes. I hope there is no regression and I think it won't be.
- PR can't be merged right now due some conflicts... I think the reason is that you used merges instead of rebase to merge branches from this PR and master to each other), which seem to create some mess in the history... I think I can cleanup this before merging this PR to latest master. But if you have a chance to cleanup history yourself a bit, it can be nice (I think the way to go is start with new branch from latest keycloak master and then cherry-pick the commits from this PR into this new branch and finally push-force this new branch to replace to the original branch of this PR).

CorneliaLahnsteiner commented:
> @CorneliaLahnsteiner @romge Thanks for all your work! From my point of view, this PR looks good now!
> 
> @pedroigor Do you have a chance to review this? I hope we can have this in KC 16, so hopefully we can address some more improvements as a follow-up tasks (EG. protocol mapper and client scopes for "acr" claim). But if you think there is any blocker, please comment.
> 
> Summary of the remaining things for this PR:
> 
> * Review of the documentation PR (I've added the review, but waiting for our documentation guy Andy if he has further comments)
> * I will try to run our internal pipeline with these changes. I hope there is no regression and I think it won't be.
> * PR can't be merged right now due some conflicts... I think the reason is that you used merges instead of rebase to merge branches from this PR and master to each other), which seem to create some mess in the history... I think I can cleanup this before merging this PR to latest master. But if you have a chance to cleanup history yourself a bit, it can be nice (I think the way to go is start with new branch from latest keycloak master and then cherry-pick the commits from this PR into this new branch and finally push-force this new branch to replace to the original branch of this PR).

@mposolda Would be great if we could have this in KC 16. And thank you for your work as well!

The only conflict is in the ConfigurationTest.java due to the changes we made for Windows (https://github.com/keycloak/keycloak/pull/7897#discussion_r735534998). I will simply reverse this change and do my best to clear the history of our branch.

@romge is currently implementing the last two integration tests:

- Check LoA mapping and how acr claim is set in the ID Token
- Check the behavior when the user is authenticated and the client requires a different LoA (step-up). This should also involve checking whether the user goes only through the missing steps to achieve a determined LoA.

If there is anything else to be done from our side, please ping us!

CorneliaLahnsteiner commented:
@mposolda When writing the integration tests, we noticed that an acr claim requested as essential claim and higher than a configured claim (e.g. requesting essential:true and an acr value of 5) no longer fails even though the LoA cannot be reached. We think this functionality has been removed from callback commits: https://github.com/keycloak/keycloak/pull/7897/commits/6e4fb3a0cd539b3eb9568c3ea7c9de67a755cac6#diff-b72231a1d2641bd3ef2dfbe97433cda0a506e6c4fafa89896222339716cee6faL305. We are not that familiar with callbacks, so we don't know exactly where to put this functionality. Could you help us to fix this?

mposolda commented:
@CorneliaLahnsteiner @romge Sorry for late response. Did you already figured this issue with "acr" with your latest changes? If not, feel free to add your test to this PR (even if the test would be failing), so I can take a look in my environment if needed.

CorneliaLahnsteiner commented:
> @CorneliaLahnsteiner @romge Sorry for late response. Did you already figured this issue with "acr" with your latest changes? If not, feel free to add your test to this PR (even if the test would be failing), so I can take a look in my environment if needed.

No. My last change fixed the behavior for unknown acr claims that are not essential. The problem I reported yesterday (https://github.com/keycloak/keycloak/pull/7897#issuecomment-960870250) concerns an unknown acr claim that is essential. The result should be a failed authentication instead of a successful authentication with the highest authentication level. We will add a test for this. @romge is in the process of completing the tests. 

To reproduce the problem, send an authentication request with an essential request and an unknown or high (e.g. 5) Acr value. After entering username/password and OTP, you should receive an authentication error. Instead, you are currently  authenticating successfully with the highest authentication level.

mposolda commented:
@CorneliaLahnsteiner @romge Sorry for the complications, but I have small update:
- I am sick and likely won't be able to finish fix the acr/essential issue by the end of this week...
- There is Keycloak 16 release planned likely this week

Based on that, I suggest if we slightly change the plan to this:
- You remove the latest commit with failing tests, so that this PR will pass all the tests
- I will merge this PR afterwards and also the docs PR, which should be already good to go AFAIK. Hopefully this can be done later today or early tomorrow. This will allow to have hopefully step-up authentication support available in Keycloak 16
- You create follow-up JIRA for the acr/essential issue and then send separate PR with the failing test. I will try to follow and figure how to fix this. IMO this is something, which can be postponed after Keycloak 16 release (acr/essential issue seems to me rather as a corner case, which will rarely happen in the production unless client are mis-configured to use invalid acr essential claim in the initial request).

WDYT?

CorneliaLahnsteiner commented:
> * You create follow-up JIRA for the acr/essential issue and then send separate PR with the failing test. I will try to follow and figure how to fix this. IMO this is something, which can be postponed after Keycloak 16 release (acr/essential issue seems to me rather as a corner case, which will rarely happen in the production unless client are mis-configured to use invalid acr essential claim in the initial request).

@mposolda.  Sorry to hear that you are sick. No problem and get well soon! 
We fully agree with your plan. In fact, the issue with essential unknown acr claims only concerns a misconfiguration, so we don't see a blocker for the PR here either. The docs PR should be good to go.
I have just removed these three failed tests. Could you approve the test runs? I can't do that because I don't have the rights. 
And I will create a follow-up JIRA for the acr/essential issue and will push the failed tests.

mposolda commented:
@CorneliaLahnsteiner Thanks! One more update: I discussed with Stian and the decision was to postpone this after Keycloak 16 to not introduce more features (and in theory potential regressions) at this stage. We will merge soon after Keycloak 16 release. Sorry for the inconvenience and Thanks again for your work on this. Hopefully we will be able to finish the journey very soon and have this in Keycloak master.

CorneliaLahnsteiner commented:
> @CorneliaLahnsteiner Thanks! One more update: I discussed with Stian and the decision was to postpone this after Keycloak 16 to not introduce more features (and in theory potential regressions) at this stage. We will merge soon after Keycloak 16 release. Sorry for the inconvenience and Thanks again for your work on this. Hopefully we will be able to finish the journey very soon and have this in Keycloak master.

@mposolda Thanks for the update. We are happy when this feature is merged, whether in KC version 16 or 17 is secondary for us.
However, the plan is to merge this PR asap after the KC 16 release, right? Because then, I would suggest not to add the three failing tests to this PR again, but to push them in the course of a new PR (as you suggested today)? Or would you rather we add the three tests here again?

mposolda commented:
> @mposolda Thanks for the update. We are happy when this feature is merged, whether in KC version 16 or 17 is secondary for us. However, the plan is to merge this PR asap after the KC 16 release, right? Because then, I would suggest not to add the three failing tests to this PR again, but to push them in the course of a new PR (as you suggested today)? Or would you rather we add the three tests here again?
> 
> BTW: Could you do a test run again? The tests failed because of a timeout (the last run was successful before the three tests were removed, so it's probably not the code causing the timeout).

@CorneliaLahnsteiner yes, I agree with NOT adding the 3 failing tests to this PR for now and just send the PR later once this one PR is merged.

However there is the issue that once I re-run the tests, I can see that LevelOfAssurance test is failing for "undertow-map" . See the failing test details here - https://github.com/keycloak/keycloak/runs/4225877555?check_suite_focus=true . Do you please have a chance to fix this test for undertow-map? Maybe it will help to just rebase this PR (again) on top of latest master, but I afraid it can be something also something else...

tnorimat commented on services/src/main/java/org/keycloak/authentication/authenticators/conditional/ConditionalLoaAuthenticator.java:
```diff
@@ -0,0 +1,75 @@
+package org.keycloak.authentication.authenticators.conditional;
```
It might be good if you could add header comments the same as other newly added source files like AuthenticatorUtil.java .

tnorimat commented on services/src/main/java/org/keycloak/authentication/authenticators/conditional/ConditionalLoaAuthenticatorFactory.java:
```diff
@@ -0,0 +1,89 @@
+package org.keycloak.authentication.authenticators.conditional;
```
It might be good if you could add header comments the same as other newly added source files like AuthenticatorUtil.java .

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/utils/AcrUtils.java:
```diff
@@ -0,0 +1,88 @@
+package org.keycloak.protocol.oidc.utils;
```
It might be good if you could add header comments the same as other newly added source files like AuthenticatorUtil.java .

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/LevelOfAssuranceFlowTest.java:
```diff
@@ -0,0 +1,272 @@
+package org.keycloak.testsuite.forms;
```
It might be good if you could add header comments the same as other newly added source files like AuthenticatorUtil.java .

tnorimat reviewed (commented):
@CorneliaLahnsteiner, @romge, I've added some review comments about header comments on newly added source files. Could you check them?

CorneliaLahnsteiner commented:
> > @mposolda Thanks for the update. We are happy when this feature is merged, whether in KC version 16 or 17 is secondary for us. However, the plan is to merge this PR asap after the KC 16 release, right? Because then, I would suggest not to add the three failing tests to this PR again, but to push them in the course of a new PR (as you suggested today)? Or would you rather we add the three tests here again?
> > BTW: Could you do a test run again? The tests failed because of a timeout (the last run was successful before the three tests were removed, so it's probably not the code causing the timeout).
> 
> @CorneliaLahnsteiner yes, I agree with NOT adding the 3 failing tests to this PR for now and just send the PR later once this one PR is merged.
> 
> However there is the issue that once I re-run the tests, I can see that LevelOfAssurance test is failing for "undertow-map" . See the failing test details here - https://github.com/keycloak/keycloak/runs/4225877555?check_suite_focus=true . Do you please have a chance to fix this test for undertow-map? Maybe it will help to just rebase this PR (again) on top of latest master, but I afraid it can be something also something else...

Hi @mposolda. I'm glad you're well again! I have just rebased this PR on top of latest master. Could you approve a new test run? It is really strange because before the three failed tests were removed, all tests were successful (except for the three failed tests). If the new test run is not successful, we will have to take a closer look.

mposolda commented:
@CorneliaLahnsteiner Unfortunately the tests for "undertow-map" still seems to fail :-( This is strange that it started to fail now after the rebase, however looks like it is a regression of this PR, which needs to be somehow addressed. I hope this will help you to figure how to run "undertow-map" tests https://github.com/keycloak/keycloak/blob/main/.github/workflows/ci.yml as it is a chance that this is passing with the standard setup, but fails just with the "undertow-map" .

Also note review comments from @tnorimat about headers and copyrights, which I missed in the initial review (Hope it is not huge issue for you to add those).

Thanks for the great work and patience for this. Really hope we are close with this PR...

CorneliaLahnsteiner commented:
> @CorneliaLahnsteiner Unfortunately the tests for "undertow-map" still seems to fail :-( This is strange that it started to fail now after the rebase, however looks like it is a regression of this PR, which needs to be somehow addressed. I hope this will help you to figure how to run "undertow-map" tests https://github.com/keycloak/keycloak/blob/main/.github/workflows/ci.yml as it is a chance that this is passing with the standard setup, but fails just with the "undertow-map" .
> 
> Also note review comments from @tnorimat about headers and copyrights, which I missed in the initial review (Hope it is not huge issue for you to add those).
> 
> Thanks for the great work and patience for this. Really hope we are close with this PR...

@mposolda. I fixed the headers and coyprights. And I was able to reproduce the behavior locally by executing group 3 of the undertow-map tests. The good news: It is not a regression of this PR. The bad news: I don't know how to fix it. The tests time out when they try to load a page completely before continuing. We used the wait.until() method to wait till the page is fully loaded. However, I looked at similar tests and they do not use the wait.until() method, but only ask if the page is present (with assertCurrent()). I don't know what is wrong about using wait.until(), but I have adjusted our test to the assertCurrent() method and locally they ran through successfully. I pushed the changes. Hopefully this fixes the tests, otherwise I have no idea how to fix them.. Can we do another test run?

mposolda commented:
@CorneliaLahnsteiner Thanks for adding copyrights and headers! Regarding failing test, I see it is still failing and unfortunately I afraid that it is a regression of this PR as LevelOfAssuranceFlowTest was added in this PR and it is the only failing test for undertow-map https://github.com/keycloak/keycloak/runs/4261773321?check_suite_focus=true ... :(

mposolda commented:
@CorneliaLahnsteiner Few tips: 
- it will be good if you add annotation `@AuthServerContainerExclude(REMOTE)` to some methods (or maybe rather whole class ) of LEvelOfAssuranceFlowTest. This may be needed as I think that setup of authentication flow with flow utils does not work with authServerRemote (For example see BrowserFlowTest for the inspiration)
- I am not sure if the `FluentWait`, which you are using in the LevelOfAssuranceFlowTest is really needed. Will it be possible to remove this custom wait utilities? For example you can take a look again at BrowserFlowTest, which does some setup of authentication flows with FlowUtils and does not use these custom wait mechanisms.

WDYT? Is it fine for you to adjust the test to same/similar pattern like BrowserFlowTest is using (since we know that BrowserFlowTest works?). If it does not help, I may try to take a look at this after Keycloak 16 release. As unfortunately we cannot merge this PR with failing test introduced in one of the groups :(

CorneliaLahnsteiner commented:
> @CorneliaLahnsteiner Few tips:
> 
> * it will be good if you add annotation `@AuthServerContainerExclude(REMOTE)` to some methods (or maybe rather whole class ) of LEvelOfAssuranceFlowTest. This may be needed as I think that setup of authentication flow with flow utils does not work with authServerRemote (For example see BrowserFlowTest for the inspiration)
> * I am not sure if the `FluentWait`, which you are using in the LevelOfAssuranceFlowTest is really needed. Will it be possible to remove this custom wait utilities? For example you can take a look again at BrowserFlowTest, which does some setup of authentication flows with FlowUtils and does not use these custom wait mechanisms.
> 
> WDYT? Is it fine for you to adjust the test to same/similar pattern like BrowserFlowTest is using (since we know that BrowserFlowTest works?). If it does not help, I may try to take a look at this after Keycloak 16 release. As unfortunately we cannot merge this PR with failing test introduced in one of the groups :(

@mposolda Thanks for approving a new test run. Sadly, it didn't work. I agree that the fault is in the test itself, meaning how we wrote it, and we should and cannot not merge this PR, till the tests are running stable. But I doubt it is in the logic or code of the step up authentication. But it cannot be rocket sience to fix the test. 

I debugged it 5 hours yesterday. It is really strange that it works with the `FluentWait` except of the undertow-map in group 3 (https://github.com/keycloak/keycloak/runs/4225877555?check_suite_focus=true). I have removed the `FluentWait` yesterday, and now the Model-tests and the undertow-map tests are failing. I don't get it, why our test (LevelOfAssuranceFlowTest) does need some kind of waiting till a page is completed, but others don't. However, thanks for the tips! I will try to add the annotation and will have a closer look at the `BrowserFlowTest`. I will ping you as soon as a new test run is required.

mposolda commented:
@CorneliaLahnsteiner @romge Sorry for the delay after longer pause. Keycloak 16 was released and I hope we can have this PR in the main branch soon. Do you please have a chance to rebase this PR? Maybe simple rebase will somehow help with failing test :-) If not, I will try to follow-up and look at the failing test, so we can finally have this PR in.

Commit:
KEYCLOAK-847 Add support for step up authentication

add step-up authentication test
add acr mappings in tests

KEYCLOAK-847 Fix behavior of unknown not essential acr claim

add level of assurance tests

fix acrValuesQueryParameter test

Remove tests for unknown essential acr claim

KEYCLOAK-847 Add headers and copyrights

KEYCLOAK-847 remove wait.until from tests

Original step up test

KEYCLOAK-847 Test fix

CorneliaLahnsteiner commented:
> 

@mposolda I have just rebased the branch on the latest main. 
I have been trying to fix the tests on a few evenings lately, but unfortunately have not managed it. I have tried your advice, but more tests have fallen over as a result. Further, I tried to implement the tests like in the SSOTest, as they also validate the acr value of a token, but without success.

mposolda commented:
@CorneliaLahnsteiner Thanks! Do you please have a chance to cherry-pick the latest commit from this branch https://github.com/mposolda/keycloak/commits/KEYCLOAK-847-step-up-auth-3 to this PR?

I've checked that it passes with map-storage profile (before this change it failed). The test is fixed and changed to align with other similar tests in the Keycloak. Hope this helps and we can have the tests passing after this.

CorneliaLahnsteiner commented:
> @CorneliaLahnsteiner Thanks! Do you please have a chance to cherry-pick the latest commit from this branch https://github.com/mposolda/keycloak/commits/KEYCLOAK-847-step-up-auth-3 to this PR?
> 
> I've checked that it passes with map-storage profile (before this change it failed). The test is fixed and changed to align with other similar tests in the Keycloak. Hope this helps and we can have the tests passing after this.

@mposolda Thank you for your support! Your changes look good. I have included your changes in this branch. Could you approve another test run? Thanks a lot.

mposolda commented:
@CorneliaLahnsteiner @romge Thanks a lot for all your work on this!

nxadm commented:
@mposolda I reply here because https://issues.redhat.com/browse/KEYCLOAK-847 does not allow comments any more.

> It will be good to discuss if there are any missing pieces or if current functionality is sufficient. The missing piece might be the ability for the client to specify the minimum level for this client (Jira is KEYCLOAK-19913 ). And maybe some other things based on the feedback.

I think it's important that mechanism also works for oidc clients that have a SAML backend as an Identity provider (a rather common scenario, I think).

CorneliaLahnsteiner commented:
> @CorneliaLahnsteiner @romge Thanks a lot for all your work on this!

@mposolda I am very happy that we made it. Thanks for your work and support as well! With that I wish you a merry Christmas and a happy new year!
